tx/textual/valuerenderer/tx.go

```diff
@@ -0,0 +1,224 @@
+	}
+
+	p1 := &textualpb.Part1{
+		ChainId:       textualData.SignerData.ChainId,
+		AccountNumber: textualData.SignerData.AccountNumber,
+		Sequence:      textualData.SignerData.Sequence,
+	}
+	p2 := &textualpb.Part2{
+		PublicKey: textualData.SignerData.PubKey,
+	}
+	p3 := &textualpb.Part3{
+		Message: txBody.Messages,
+		Memo:    txBody.Memo,
+		Fees:    txAuthInfo.Fee.Amount,
+	}
+	p4 := &textualpb.Part4{
+		FeePayer:   txAuthInfo.Fee.Payer,
+		FeeGranter: txAuthInfo.Fee.Granter,
+	}
+	p5 := &textualpb.Part5{}
+	if txAuthInfo.Tip != nil {
+		p5.Tip = txAuthInfo.Tip.Amount
+		p5.Tipper = txAuthInfo.Tip.Tipper
+	}
+	// Find all other tx signers than the current signer.
+	otherSigners := make([]*txv1beta1.SignerInfo, 0, len(txAuthInfo.SignerInfos)-1)
+	for _, si := range txAuthInfo.SignerInfos {
+		if bytes.Equal(si.PublicKey.Value, textualData.SignerData.PubKey.Value) {
+			continue
+		}
+
+		otherSigners = append(otherSigners, si)
+	}
+	p6 := &textualpb.Part6{
+		GasLimit:                    txAuthInfo.Fee.GasLimit,
+		TimeoutHeight:               txBody.TimeoutHeight,
+		OtherSigner:                 otherSigners,
+		ExtensionOptions:            txBody.ExtensionOptions,
+		NonCriticalExtensionOptions: txBody.NonCriticalExtensionOptions,
+		HashOfRawBytes:              getHash(textualData.BodyBytes, textualData.AuthInfoBytes),
+	}
+
+	screens1, err := vr.formatPart(ctx, p1, false)
+	if err != nil {
+		return nil, err
+	}
+	screens2, err := vr.formatPart(ctx, p2, true)
+	if err != nil {
+		return nil, err
+	}
+	screens3, err := vr.formatPart(ctx, p3, false)
+	if err != nil {
+		return nil, err
+	}
+	// Replace:
+	// "Messages: <N> Any"
+	// with:
+	// "This transaction has <N> Message"
+	screens3[0].Text = fmt.Sprintf("This transaction has %d Message", len(txBody.Messages))
+	screens4, err := vr.formatPart(ctx, p4, true)
+	if err != nil {
+		return nil, err
+	}
+	screens5, err := vr.formatPart(ctx, p5, false)
+	if err != nil {
+		return nil, err
+	}
+	screens6, err := vr.formatPart(ctx, p6, true)
+	if err != nil {
+		return nil, err
+	}
+
+	screens := append(screens1, append(screens2, append(screens3, append(screens4, append(screens5, screens6...)...)...)...)...)
+
+	return screens, nil
+}
+
+func (vr txValueRenderer) formatPart(ctx context.Context, m proto.Message, expert bool) ([]Screen, error) {
+	messageVR := NewMessageValueRenderer(vr.tr, m.ProtoReflect().Descriptor())
+	screens, err := messageVR.Format(ctx, protoreflect.ValueOf(m.ProtoReflect()))
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove 1st screen which is the message name
+	screens = screens[1:]
+
+	// Remove indentations on all subscreens
+	for i := range screens {
+		screens[i].Indent--
+		if expert {
+			screens[i].Expert = true
+		}
+	}
+
+	return screens, nil
+}
+
+// getHash gets the hash of raw bytes to be signed over:
+// HEX(sha256(len(body_bytes) ++ body_bytes ++ len(auth_info_bytes) ++ auth_info_bytes))
+func getHash(bodyBz, authInfoBz []byte) string {
```

Comment:
There's no existing function to use? Then please cross-reference to the code that computes the TxHash for other signing modes, and cross-reference there back to here, so both locations can be kept consistent.

Reply:
We're signing over custom bytes here `len(body_bytes) ++ body_bytes ++ len(auth_info_bytes) ++ auth_info_bytes`, as defined in the Textual ADR. That's why there's no other function to use.

Instead of signing over the hash of these custom bytes, we could indeed sign over the hash of the existing [SIGN_MODE_DIRECT sign doc](https://github.com/cosmos/cosmos-sdk/blob/a1dc6d999809a9f5a334edae49637c3522bf66ae/proto/cosmos/tx/v1beta1/tx.proto#L49). But we would be signing over chainId and accountNumber twice.

I have a slightly small preference to keep it like we currently do, so that Textual doesn't depend on SIGN_MODE_DIRECT, but also happy to switch if you think avoiding these custom bytes signing is a good idea.

Reply:
No, you're convinced me.